### PR TITLE
Create pybind11-based Python bindings

### PR DESCRIPTION
**Pull request by francesco-romano**
This PR introduces an alternative way to generate Python bindings.

Currently only basic classes are wrapped, but this can be extended later.

A part the usual changelog missing (will do at the end of the review), the PR is missing also the install phase. @diegoferigo let me know your input on this / where do you expect the python package to be installed.

### CODE REVIEW

**Comment by diegoferigo** [comment]
This is an amazing work @francesco-romano, awesome :) It's quite a massive coverage of most of the core features of the library. I didn't go through in detail on all the binded classes, though the tests look good. Having this alternative to SWIG for `{core,model,sensors,modelio}` is really promising. It would be nice on the long term also including `{high-level,inverse-kinematics}`, that after all are what most of downstream users use.

Being more familiar with SWIG, it seems to me that pybind11 trade off maintainability effort with flexibility, particularly in obtaining more idiomatic bindings by default and memory management. There seem to be also [native support to NumPy](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html), not included in this PR but that could be interesting to explore in the future.

Now, how we should handle duplicate Python bindings is open to discussion cc @traversaro. Despite I spent time and effort in improving the SWIG bindings, **I don't hide that offering a more flexible solution based on pybind11 is appealing**, especially with the great start given by this PR. However, with the SWIG improvements we managed to minimize downstream API changes requiring only a single-line import statement update. Here we're talking on completely different APIs (and I feel personally affected since I have Python projects based on the existing SWIG bindings). Just to know, do we have any schedule for the v2 release?

**About naming**, I saw you used the same fs pattern `idyntree.bindings` of SWIG. Since the two bindings provide different APIs, they cannot share the name. However, I think the previous discussion point should be addressed first.

**About the install step**, I would keep it disabled for the time being. If the new SWIG bindings are still experimental, these new pybind11 are even more than that :) Mainly because there's not yet the same C++ coverage and NumPy support is still missing. As soon as we decide how to deal with the double system, we can finalize it.

I conclude saying that, in order to maintain the current compatibility and avoid breaking changes, **both bindings could co-exists** inside the new `idyntree` package. As I mentioned, only one can be imported as `import idyntree.bindings` and we should decide what's the horse we want to bet on in the long term [1].

---

Disclaimer: I'm not familiar with pybind11.

[1] Of course we could also do a `import idyntree.swig.bindings` and `import idyntree.pybind.bindings`, but I'd prefer having a more compact `import idyntree.bindings` for the mostly supported and `import idyntree.<othername>` for the other one.

**Comment by francesco-romano** [comment]
Added also install phase with support to RPATH.
Note that, this is only for the pybind11 bindings.
RPATH support will be "for free" also for the SWIG bindings but only in `devel` as the installation path is changed there.

Reply to @diegoferigo in a separate message

**Comment by francesco-romano** [comment]
Thanks @diegoferigo 
Yes, `pybind11` has this trade-off. There seems to be also automatic tools, but I was more familiar with native `pybind11` so this is why I did this.
While at the beginning I thought was a huge disadvantage, I discover that is has its pros to select which API you want to expose to Python (and some of these can change to make the bindings more similar to Python than C++).

> It would be nice on the long term also including `{high-level,inverse-kinematics}`, that after all are what most of downstream users use.

Sure. I currently needed the core, model libraries, but given the skeleton it should be trivial to add support to additional iDynTree libraries.

> There seem to be also [native support to NumPy](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html), not included in this PR but that could be interesting to explore in the future.

I have not tested all NumPy support (as it is probably more useful for `high-level`) but basic support is there for `Vector` and `Matrix`. You can create a numpy array wrapping (or copying) the iDynTree type. There should be also a test showing this.

> Now, how we should handle duplicate Python bindings is open to discussion cc @traversaro. Despite I spent time and effort in improving the SWIG bindings, **I don't hide that offering a more flexible solution based on pybind11 is appealing**, especially with the great start given by this PR. However, with the SWIG improvements we managed to minimize downstream API changes requiring only a single-line import statement update. Here we're talking on completely different APIs (and I feel personally affected since I have Python projects based on the existing SWIG bindings). Just to know, do we have any schedule for the v2 release?

As discussed with @traversaro I will try to rebase this on `devel` and see what happens.

The API is different w.r.t. SWIG for an explicit choice: the SWIG one is too much C++ and here I intentionally hide the fact that these are bindings (as much as I could). 

> 
> **About naming**, I saw you used the same fs pattern `idyntree.bindings` of SWIG. Since the two bindings provide different APIs, they cannot share the name. However, I think the previous discussion point should be addressed first.

I think it would be strange to have both. What about adding a check in the CMake that only one between SWIG and pybind11 can be compiled? 

> **About the install step**, I would keep it disabled for the time being. If the new SWIG bindings are still experimental, these new pybind11 are even more than that :) Mainly because there's not yet the same C++ coverage and NumPy support is still missing. As soon as we decide how to deal with the double system, we can finalize it.
> 

I actually added the install step. Cleaning up also the RPATH support that was broken even with the native bindings.
I do not see why we should not provide the installation. I am allergic to build directories. But in the end this is up to you.

> I conclude saying that, in order to maintain the current compatibility and avoid breaking changes, **both bindings could co-exists** inside the new `idyntree` package. As I mentioned, only one can be imported as `import idyntree.bindings` and we should decide what's the horse we want to bet on in the long term [1].
> 
I would say that they should not co-exist in a user build/install space.
But I would keep the same import name just to avoid bad/explicit bindings type in the import.
Note that while the install discussion does not influence me, this has an impact also on my workflow. Ideally I would like to keep my code as much as possible compatible with other Python code using pybind11 bindings (e.g. tests)

**Comment by francesco-romano** [comment]
@traversaro rebased to `devel`

**Comment by diegoferigo** [comment]
>> **About naming**, I saw you used the same fs pattern `idyntree.bindings` of SWIG. Since the two bindings provide different APIs, they cannot share the name. However, I think the previous discussion point should be addressed first.
>
> The API is different w.r.t. SWIG for an explicit choice: the SWIG one is too much C++ and here I intentionally hide the fact that these are bindings (as much as I could).

This is great from the user perspective, being able with pybind to work on a lower level wrt SWIG shows its advantages here (it can be done automatically also with SWIG, but in the previous PRs I decided not to break user code beyond the import statement). 

>> About the install step, I would keep it disabled for the time being. If the new SWIG bindings are still experimental, these new pybind11 are even more than that :) Mainly because there's not yet the same C++ coverage and NumPy support is still missing. As soon as we decide how to deal with the double system, we can finalize it.
>
> I actually added the install step. Cleaning up also the RPATH support that was broken even with the native bindings.
I do not see why we should not provide the installation. I am allergic to build directories. But in the end this is up to you.

Cool, thanks also for fixing RPATH, we had some issues in the past with the bindings and no one tried to fix it :) I'm not saying that we should not provide installation, what I'm not sure is how we should distribute the two bindings. This is the last open question, I provide some more details below.

>> I conclude saying that, in order to maintain the current compatibility and avoid breaking changes, **both bindings could co-exists** inside the new `idyntree` package. As I mentioned, only one can be imported as `import idyntree.bindings` and we should decide what's the horse we want to bet on in the long term [1].
>
> I would say that they should not co-exist in a user build/install space.
But I would keep the same import name just to avoid bad/explicit bindings type in the import.
Note that while the install discussion does not influence me, this has an impact also on my workflow. Ideally I would like to keep my code as much as possible compatible with other Python code using pybind11 bindings (e.g. tests)

I disagree on preventing the co-existence of the bindings. If we support both systems, why not allowing users to use them on different projects (or parts of the same project) without reinstalling entirely iDynTree? The key here is using different import statements, and this means changing this PR since it installs the pybind bindings as `idyntree.bindings` as for the native bindings.

What I propose, and it's open to discussion, is installing the bindings as follows:

```
idyntree/
|-- __init__.py
|-- bindings_pybind.py
|-- bindings_swig.py
|-- libpybind.so
`-- libswig.so
```

With the following `__init__.py`:

```python
# Note: I don't like much these names, suggestions are welcome
from . import bindings_pybind
from . import bindings_swig

from . import bindings_swig as bindings
```

This way, users can use both bindings by importing the specific module, or using the default one that can be imported with `import idyntree.bindings`. This allows to switch from one to another in the future (better if between mayor releases) while allowing legacy code to run with just changing the import statement. In this way, we could proceed with SWIG since it's now the most complete system and, if we like, progressively migrate to pybind as soon as it will become as complete as SWIG.

**Comment by francesco-romano** [comment]
As discussed offline, we now install both bindings with different names, i.e. `swig` and `pybind` for `SWIG`- and `pybind`-based bindings respectively.

SWIG bindings are also imported as `bindings`.

**Review by traversaro — APPROVED** [review]

File: .github/workflows/python.yml [hunk 1]
```diff
@@ -42,7 +42,7 @@ jobs:
         run: pip -v install dist/*.tar.gz
 
       - name: Test import
-        run: python -c 'import idyntree.bindings'
+        run: python -c 'import idyntree.swig'
```
Comment (traversaro) [inline .github/workflows/python.yml:45]: To test that import.bindings works fine (it should import swig by defaults, unless I am did not get the discussion in the PR), perhaps we should leave the test as it is? 
```suggestion
        run: python -c 'import idyntree.bindings'
```

File: .github/workflows/python.yml [hunk 2]
```diff
@@ -56,7 +56,7 @@ jobs:
         run: pip -v install dist/*.whl
 
       - name: Test import
-        run: python -c 'import idyntree.bindings'
+        run: python -c 'import idyntree.swig'
```
Comment (traversaro) [inline .github/workflows/python.yml:59]:
```suggestion
        run: python -c 'import idyntree.bindings'
```
See previous comment.

Comment (francesco-romano) [reply .github/workflows/python.yml]: Done

File: bindings/CMakeLists.txt [hunk 3]
```diff
@@ -11,6 +11,14 @@ option(IDYNTREE_USES_OCTAVE "Do you want to create the OCTAVE bindings" FALSE)
 option(IDYNTREE_GENERATE_MATLAB "Enable if you have the experimental version of SWIG necessary for generating the Matlab wrapper" FALSE)
 
 find_package(SWIG)
+
+# Setup installation directory for Python bindings.
+if(IDYNTREE_PACKAGE_FOR_PYPI)
+    set(PYTHON_INSTDIR ${CMAKE_INSTALL_PREFIX})
+else()
+    set(PYTHON_INSTDIR ${IDYNTREE_PYTHON_INSTALL_DIR}/idyntree)
```
Comment (diegoferigo) [inline bindings/CMakeLists.txt:19]: While configuring the project, without enabling Python support, I get:

```
CMake Warning (dev) at bindings/CMakeLists.txt:19 (set):
  uninitialized variable 'IDYNTREE_PYTHON_INSTALL_DIR'
```

File: bindings/CMakeLists.txt [hunk 4]
```diff
@@ -67,3 +75,41 @@ if(IDYNTREE_USES_PYTHON OR
...
+        file(APPEND "${CMAKE_CURRENT_BINARY_DIR}/__init__.py" "from . import swig${NEW_LINE}")
+        # Create also an alias from swig to the old `bindings` import.
+        file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/bindings.py" "from .swig import *${NEW_LINE}")
+        install(FILES "${CMAKE_CURRENT_BINARY_DIR}/bindings.py"
+                DESTINATION ${PYTHON_INSTDIR})
+        # Add the alias in the __init__.py
+        file(APPEND "${CMAKE_CURRENT_BINARY_DIR}/__init__.py" "from . import bindings${NEW_LINE}")
+    endif()
+    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/__init__.py"
+             DESTINATION ${PYTHON_INSTDIR})
```
Comment (diegoferigo) [inline bindings/CMakeLists.txt:114]: This is not exactly what I had in mind but I just realized that the import could not be done as I originally thought.

### What I was thinking

1. Remove the `bindings.py` file
2. Obtain the following `__init__.py` file:

```py
from . import pybind
from . import swig

from . import swig as bindings
```

This, however, does not allow downstream projects to use `import idyntree.bindings` but only `from idyntree import bindings`.

### What we have now

Now the init files and import statements are a bit intertwined and not super clean, but they enable using both import types. I can't see any problem that could emerge from it, and from my side it's ok.

**Review by diegoferigo — APPROVED** [review]
It looks great, just the last comments

File: bindings/CMakeLists.txt (hunk 3 above)
Comment (francesco-romano) [reply bindings/CMakeLists.txt]: This should be solved now.

File: README.md [hunk 5]
```diff
@@ -148,6 +148,33 @@ You should add to the `PYTHONPATH` enviromental variable the install path of the
...
+```python
+disutils.sysconfig.get_python_lib(1,0,prefix='')
+```
+
+for example: `lib/python3.8/site-packages`.
+
+Finally to use the bindings in your Python code, simply import the package:
+
+```python
+import idyntree.bindings as iDynTree
```
[the remark this answers is not in the document]
Comment (francesco-romano) [reply README.md]: Done.

File: bindings/CMakeLists.txt (hunk 4 above)
Comment (francesco-romano) [reply bindings/CMakeLists.txt:116]: Yes. It is exactly as you said. If you want to have the alias and `import idyntree.bindings` you need to have the Python `bindings.py` file.

**Comment by traversaro** [comment]
Thanks a lot @francesco-romano !

**Pull request merged**